### PR TITLE
add rwlock in the interfaces of index

### PR DESCRIPTION
**Pull request by wxyucs**

### CODE REVIEW

File: src/index/diskann.cpp [hunk 1]
```diff
@@ -303,6 +305,7 @@ DiskANN::knn_search(const DatasetPtr& query,
             return invalid->Test(bit_index);
         };
     }
+    std::shared_lock lock(rw_mutex_);
```
Comment (jiaweizone) [inline src/index/diskann.cpp]: Duplicate lock with line 372?

Comment (wxyucs) [reply src/index/diskann.cpp]: removed the dup lock

File: src/index/hnsw.cpp [hunk 2]
```diff
@@ -242,6 +246,7 @@ HNSW::knn_search(const DatasetPtr& query,
         std::priority_queue<std::pair<float, size_t>> results;
         double time_cost;
         try {
+            std::shared_lock lock(rw_mutex_);
```
Comment (jiaweizone) [inline src/index/hnsw.cpp]: Does L239 have thread-safe issue ?

File: src/index/hnsw.cpp [hunk 3]
```diff
@@ -209,6 +213,8 @@ HNSW::knn_search(const DatasetPtr& query,
         CHECK_ARGUMENT(k > 0, fmt::format("k({}) must be greater than 0", k))
         k = std::min(k, GetNumElements());
 
+        std::shared_lock lock(rw_mutex_);
+
         // check search parameters
         auto params = HnswSearchParameters::FromJson(parameters);
         alg_hnsw->setEf(std::max(params.ef_search, k));
```
Comment (jiaweizone) [inline src/index/hnsw.cpp]: In 'shared_lock' scope here  have thread-safe issue, move parameter 'ef' to stack variable ?

File: src/index/hnsw.cpp [hunk 4]
```diff
@@ -330,6 +337,7 @@ HNSW::range_search(const DatasetPtr& query,
         std::priority_queue<std::pair<float, size_t>> results;
         double time_cost;
         try {
+            std::shared_lock lock(rw_mutex_);
```
Comment (jiaweizone) [inline src/index/hnsw.cpp:339]: L334 have thread-safe issues

File: src/index/hnsw.cpp [hunk 5]
```diff
@@ -585,6 +599,7 @@ HNSW::feedback(const DatasetPtr& query,
         return 0;
     }
 
+    std::shared_lock lock(rw_mutex_);
```
Comment (jiaweizone) [inline src/index/hnsw.cpp]: thread-safe issues: L616 this->feedback maybe update  'conjugate_graph_', we need upgrade to unique lock ?

File: tests/test_multi_thread.cpp [hunk 6]
```diff
@@ -234,3 +243,70 @@ TEST_CASE("HNSW Multi-threading", "[ft][hnsw]") {
...
+                auto result = index->KnnSearch(query, 2, str_parameters);
+                return result.has_value();
+            }));
+    }
+    for (auto& res : insert_results) {
+        REQUIRE(res.get() == 0);
+    }
+
+    for (int i = 0; i < search_results.size(); ++i) {
+        REQUIRE(search_results[i].get());
```
Comment (inabao) [inline tests/test_multi_thread.cpp:310]: Why not verify the correctness of the results?

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**